### ISSUE TITLE
time series to image on gpu? 

### ISSUE THREAD

**Issue by mohitsawwalakhe** [timeseriesAI/tsai#492]
`

> from pyts.image.gaf import GramianAngularField

from pyts.image import MarkovTransitionField, RecurrencePlotclass 

TSToGADF(Transform):
    r"""Transforms a time series batch to a 4d TSImage (bs, n_vars, size, size) by applying Gramian Angular Difference Field.
    It requires either input to be previously normalized between -1 and 1 or set range to (-1, 1)"""
    order = 98

    def __init__(self, size=224, cmap=None, range=None, **kwargs):
        self.size,self.cmap,self.range = size,cmap,range
        self.encoder = GramianAngularField(image_size=1., sample_range=self.range, method='d', **kwargs)

    def encodes(self, o: TSTensor):
        bs, *_, seq_len = o.shape
        size = ifnone(self.size, seq_len)
        if size != seq_len:
            o = F.interpolate(o.reshape(-1, 1, seq_len), size=size, mode='linear', align_corners=False)[:, 0]
        else:
            o = o.reshape(-1, seq_len)

>         output = self.encoder.fit_transform(o.cpu().numpy()).reshape(bs, -1, size, size) / 2 + .5

        if self.cmap and output.shape[1] == 1:
            output = TSImage(plt.get_cmap(self.cmap)(output)[..., :3]).squeeze(1).permute(0,3,1,2)
        else: output = TSImage(output)
 

>        return output.to(device=o.device) 

`

So, As far as my understanding, the 2nd quoted code line is utilising CPU for generating GADF of an image and then is shifted to another device(3nd quoted code line)
can this operation be performed on a GPU instead?
This implementation is common in every ts to image class, except tstoplot.
The implementation of functions to perform ts to img are being imported from pyts library(1st quoted line of code), they seem to be working with numba.
Advantage - this would significantly increase the speed of the entire pipeline(image generation and training), and possibly less RAM consumption.

**Comment by oguiza** [comment]
Hi @mohitsawwalakhe, thanks for your interest in tsai.
Unfortunately the pyts library doesn't work on gpu. It seems to use numba, so I'm sure there's some acceleration there, but it might potentially be faster if it worked with data on the gpu. 
But I'm afraid there's nothing I can do about that. 
You may want to raise this question in the pyts GitHub repo.

**Comment by mohitsawwalakhe** [comment]
can we re-implement the said code in pyts in PyTorch for Tsai, I am new to open source and i will try to reimplement the pyts code in torch for GPU compatibility.

**Comment by oguiza** [comment]
I don't have time to work on this initiative at the moment. This is not a core part of the `tsai` library. That's why we added pyts as an optional package. 
You could get in touch with the pyts core developers and ask them if they are interested in your initiative. I think it'd be great if pyts added gpu support.

**Comment by mohitsawwalakhe** [comment]
Yes, i will raise this issue with the pyts team. Another solution - is since a time series is being converted to the image on the fly( This was mentioned in the tutorial notebook [Imaging time series]). They can be converted all at once to images, and then can be moved to GPU afterwards, this would possibly speed up the training process.

**Comment by oguiza** [comment]
@mohitsawwalakhe, is it ok if I close this issue?

**Comment by mohitsawwalakhe** [comment]
@oguiza yes, thank you for helping me.
